replication/AmbryReplication.tla

```diff
@@ -0,0 +1,328 @@
+
+(* Receive a PUT request with success target of 2.
+ * For the request to succeed, atleast two replicas must be up to accept the write.
+ *)
+RecvPut ==
+    /\ \E r1, r2 \in Replicas:
+        /\ r1 # r2
+        /\ blobIdCtr < NumBlobs
+        /\ blobIdCtr' = blobIdCtr + 1
+        /\ replicaStates[r1] = CanTakeWrite
+        /\ replicaStates[r2] = CanTakeWrite
+        /\ [blobid |-> blobIdCtr, blobmessage |-> PUT ] \notin storedMessagesPerReplica[r1]
+        /\ [blobid |-> blobIdCtr, blobmessage |-> PUT ] \notin storedMessagesPerReplica[r2]
+        /\ storedMessagesPerReplica' = [r \in DOMAIN storedMessagesPerReplica |-> 
+            IF ((r = r1) \/ (r = r2)) THEN
+                storedMessagesPerReplica[r] \union { [blobid |-> blobIdCtr, blobmessage |-> PUT ] }
+            ELSE
+                storedMessagesPerReplica[r]
+            ]
+    /\ UNCHANGED <<replicaStates, inflightReplicationMessages, deleteCtr, updateTtlCtr, inflightReplicationRequest, inflightReplicationResponse, recvReplicationReqSet, sendRecplicationSet, recvReplicationRespSet>>
+
+(* Receive a DELETE request with success target of 2.
+ * For the request to succeed, atleast two replicas
+ *    - must be up AND
+ *    - must have a PUT record for the blob.
+ *)
+RecvDelete(blobId) ==
+    /\ \E r1, r2 \in Replicas:
+        /\ r1 # r2
+        /\ blobId < blobIdCtr
+        /\ replicaStates[r1] = CanTakeWrite
+        /\ replicaStates[r2] = CanTakeWrite
+        /\ deleteCtr' = deleteCtr + 1
+        /\ [blobid |-> blobId, blobmessage |-> PUT ] \in storedMessagesPerReplica[r1]
+        /\ [blobid |-> blobId, blobmessage |-> PUT ] \in storedMessagesPerReplica[r2]
+        /\ [blobid |-> blobId, blobmessage |-> DELETE ] \notin storedMessagesPerReplica[r2]
+        /\ [blobid |-> blobId, blobmessage |-> DELETE ] \notin storedMessagesPerReplica[r1]
+        /\ storedMessagesPerReplica' = [rr \in DOMAIN storedMessagesPerReplica |-> 
+            IF ((rr = r1) \/ (rr = r2)) THEN
+                storedMessagesPerReplica[rr] \union { [ blobid |-> blobId, blobmessage |-> DELETE ] }
+            ELSE
+                storedMessagesPerReplica[rr]
+            ]
+    /\ UNCHANGED <<replicaStates, blobIdCtr, inflightReplicationMessages, updateTtlCtr, inflightReplicationRequest, inflightReplicationResponse, recvReplicationReqSet, sendRecplicationSet, recvReplicationRespSet>>
+
+(* Receive a UPDATE TTL request with success target of 2.
+ * For the request to succeed, atleast two replicas
+ *    - must be up AND
+ *    - must have a PUT record for the blob AND
+ *    - must not have a DELETE record for the blob.
+ *)
+RecvUpdateTtl(blobId) ==
+    /\ \E r1, r2 \in Replicas:
+        /\ r1 # r2
+        /\ blobId < blobIdCtr
+        /\ replicaStates[r1] = CanTakeWrite
+        /\ replicaStates[r2] = CanTakeWrite
+        /\ [ blobid |-> blobId, blobmessage |-> PUT ] \in storedMessagesPerReplica[r1]
```

Comment:
There seems to be an implication of an ordering requirement for receiving blob operations.  Is this true to life?

Reply:
Actually I take that back.  Theres only a requirement of creation and deletion must start and terminate all update operations on a blob.  Is there any issue with updates that don't get fully applied because delete is in the mix?  Do we care?